### PR TITLE
SocketAsyncEventArgs should not be shared between parallel *Async operations

### PR DESCRIPTION
**Pull request by dimhotepus**
Caching `SocketAsyncEventArgs` and reusing them causes sporadic socket exceptions in `SendAsync` / `ReceiveAsync`.

`SocketAsyncEventArgs` instances expected to be unique for each `*Async` call performed in parallel.

### CODE REVIEW

**Comment by juliusfriedman** [comment]
Thanks for your PR.

I don't think this is the `caching and reusing` causing the error, it likely stems from the design of the `ClientSession` which used to use `BeginSend` and `EndSend` but which no longer work correctly for .net core and I have since moved on to use `SocketAsyncEventArgs`.

Long story short this occurs because I am using a `SocketAsyncEventArgs` per operation rather than a single one per Socket.

It can be remedied by using a single `SocketAsyncEventArgs` for both send and receive operations but would require switching on the logic depending on which operation the `SocketAsyncEventArgs` was used for.

Please let me know where this is causing problems for you and I will look into it otherwise I will advise this is a caveat and by design for now.

I will try to measure performance with your changes e.g. without reusing the `SocketAsyncEventArgs` vs having the exception, hopefully I will find that your change benefits the library and if so I will happily accept it. Until then please be patient and let me know where your having issues so I can try to replicate it and look into it.

Thanks again for your contribution!

**Comment by dimhotepus** [comment]
> Thanks for your PR.
> 
> I don't think this is the `caching and reusing` causing the error, it likely stems from the design of the `ClientSession` which used to use `BeginSend` and `EndSend` but which no longer work correctly for .net core and I have since moved on to use `SocketAsyncEventArgs`.
> 
> Long story short this occurs because I am using a `SocketAsyncEventArgs` per operation rather than a single one per Socket.
> 
> It can be remedied by using a single `SocketAsyncEventArgs` for both send and receive operations but would require switching on the logic depending on which operation the `SocketAsyncEventArgs` was used for.
> 
> Please let me know where this is causing problems for you and I will look into it otherwise I will advise this is a caveat and by design for now.
> 
> I will try to measure performance with your changes e.g. without reusing the `SocketAsyncEventArgs` vs having the exception, hopefully I will find that your change benefits the library and if so I will happily accept it. Until then please be patient and let me know where your having issues so I can try to replicate it and look into it.
> 
> Thanks again for your contribution!

I've used `TestServer` method from `UnitTests` as you provided in https://github.com/juliusfriedman/net7mma_core/issues/15
I've commented all media sources and added my own one with USA road camera (can open in VLC):

`server.TryAddMedia(new Media.Rtsp.Server.MediaTypes.RtspSource("R2_059", "rtsp://8.15.251.101:1935/rtplive/R2_059"));`

And  open result in VLC (`192.168.0.86` is my local IP):

`rtsp://192.168.0.86:8554/live/R2_059`

And sometimes I see in console:
```
An asynchronous socket operation is already in progress using this SocketAsyncEventArgs instance.
    at System.Net.Sockets.SocketAsyncEventArgs.ThrowForNonFreeStatus(Int32 status)
   at System.Net.Sockets.SocketAsyncEventArgs.StartOperationCommon(Socket socket, SocketAsyncOperation operation)
   at System.Net.Sockets.Socket.ReceiveAsync(SocketAsyncEventArgs e, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.ReceiveAsync(SocketAsyncEventArgs e)
   at Media.Rtsp.RtspServer.ProcessSendComplete(Object sender, SocketAsyncEventArgs e) in D:\github\net7mma_core\RtspServer\RtspServer.cs:line 2175
```

`SocketAsyncEventArgs` stores state per operation, so it should not be reused between operations.

**Comment by juliusfriedman** [comment]
There is a lot of state in the SocketAsyncEventArgs `class` however its disposable and should also be disposed of when no longer used to free that state up, allowing it to survive will cause GC pressure which causes stalls and should be avoided if possible.

Taking a look at https://github.com/rashiph/DecompliedDotNetLibraries/blob/master/System/System/Net/Sockets/SocketAsyncEventArgs.cs

I see that this is likely caused by how I use a 0 byte Send which is followed by a Receive and both are happening in a very short period.

If you change the code as follows do you still get an exception? 

Also does the exception really hurt anything as its' handled pretty quickly?

Change as follows:

```
internal SocketAsyncEventArgs socketAsyncEventArgs;
```

Then use `socketAsyncEventArgs` everywhere instead of separate send and receive.

You should find that the exception also goes away but this causes more refactoring to need to be done on the server level as you have to switch on the operation to determine if this is a send or receive on the callback.

I choose to design it as I did for now for the future because its possible that just as `BeginSend` and `EndSend` no longer work, `SendAsync` may work differently in the future.

See example for their pool here:

https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socketasynceventargs?view=net-7.0

They re-use them also, just not per operation and don't seem to have the exception which I have (its because I use separate dedicated `SocketAsyncEventArgs ` for send and receive)

Let me know if this is making any sense and what I can do to help.

File: RtspServer/ClientSession.cs [hunk 1]
```diff
@@ -393,16 +390,15 @@ public void StartReceive()
 
             if (object.ReferenceEquals(m_RtspSocket, null).Equals(false) && HasRuningServer)
             {
-                if (LastRecieve == null)
+                var receiveArgs = new SocketAsyncEventArgs
```
Comment (juliusfriedman) [inline RtspServer/ClientSession.cs:393]: If this really fixed the issue for you then try to just use a single SocketAsyncEventArgs which is disposed when the Session is disposed, you might have to refactor the server or client session to have a jump method which can switch out based on SocketAsyncEventArgs.SocketOperation to determine if was a send or receieve.

**Comment by juliusfriedman** [comment]
I played around with having only a single SocketAsyncEventArgs and it didn't help the problem, I still encountered it so at this time I am more inclined to leave it as it is.

```
        internal SocketAsyncEventArgs SocketAsyncEventArgs;
        
        ClientSession(){
        
        SocketAsyncEventArgs = new ();
            SocketAsyncEventArgs.Completed += m_Server.ProcessSocketAsyncEventArgs;
            SocketAsyncEventArgs.UserToken = this;
            
            }
            
            
            StartReceive(){
            
            if (object.ReferenceEquals(m_RtspSocket, null).Equals(false) && HasRuningServer)
            {
                SocketAsyncEventArgs.SetBuffer(m_Buffer.Array, m_Buffer.Offset, m_Buffer.Count);
                SocketAsyncEventArgs.RemoteEndPoint = RemoteEndPoint;
                if (!m_RtspSocket.ReceiveAsync(SocketAsyncEventArgs))
                    ThreadPool.QueueUserWorkItem((o) => m_Server.ProcessSocketAsyncEventArgs(m_Server, SocketAsyncEventArgs));
            }
            }
            
            SendRtspData(){
SocketAsyncEventArgs.RemoteEndPoint = other ?? RemoteEndPoint;
                SocketAsyncEventArgs.SocketFlags = flags;
                SocketAsyncEventArgs.SetBuffer(m_SendBuffer, offset, length);
                if (!m_RtspSocket.SendAsync(SocketAsyncEventArgs))
                    ThreadPool.QueueUserWorkItem((o) => m_Server.ProcessSocketAsyncEventArgs(m_Server, SocketAsyncEventArgs));
}
        
```

RtspServer

```
internal void ProcessSocketAsyncEventArgs(object sender, SocketAsyncEventArgs e)
        {
            switch (e.LastOperation)
            {
                default:
                case SocketAsyncOperation.Receive: ProcessReceive(sender, e); break;
                case SocketAsyncOperation.Send: ProcessSendComplete(sender, e); break;
            }
        }
        ```

**Comment by juliusfriedman** [comment]
This is definitely caused by me having a Receive already being processed, I will look into a better way to resolve this but its currently handled appropriately and logged.

**Pull request closed without merging**